### PR TITLE
refactoring

### PR DESCRIPTION
**Pull request by kbhonagiri16**
refactoring codebase

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+// Expired Error
+type Expired struct{}
```
Comment (zaletniy) [inline client.go]: Please add Error. LIke TokenExpiredError

**Comment by kbhonagiri16** [comment]
@zaletniy illia please review

File: client.go [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var emailValid = regexp.MustCompile(`^[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,4}$`)
```
Comment (zaletniy) [inline client.go]: It is not client's responsibility to validate it. Server should do it if needed.

File: client.go [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+// InvalidEmail Error
+type InvalidEmail struct{}
```
Comment (zaletniy) [inline client.go]: Just remove

File: client.go [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+func (e LoginMissing) Error() string {
+	return "Missing Login Parameter"
+}
+
+// PasswordMissing Error
+type PasswordMissing struct{}
+
+func (e PasswordMissing) Error() string {
+	return "Missing Password Parameter"
+}
```
Comment (zaletniy) [inline client.go]: All these missing things should be validated on server.
Please create one error type for all HTTP 422 responces and just copy error message from server to error object model.

Remove all these *Missing

File: client.go [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+// NotInteger Error
+type NotInteger struct{}
+
+func (e NotInteger) Error() string {
+	return "ID provided is not Integer"
+}
```
Comment (zaletniy) [inline client.go]: As *Missing

File: client.go [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+	OrgID    string `json:"orgID"`
+	UserID   string `json:"userID"`
+	Login    string `json:"login"`
+	Role     string `json:"role"`
+	Email    string `json:"email"`
+	Password string `json:"password"`
+}
+
+// OrgList Get organization list
+type OrgList struct {
```
Comment (zaletniy) [inline client.go]: Just Org
It becomes List with []

File: client.go [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+
+// Token for AuthToken
+type Token struct {
+	OrganizationID string    `json:"organizationId"`
+	ExpiresAt      time.Time `json:"expiresAt"`
+	IsAdmin        bool      `json:"isAdmin"`
+}
+
+// OrgUserList Get Users in organization
+type OrgUserList struct {
```
Comment (zaletniy) [inline client.go]: UserInOrganization according to spec

File: client.go [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+
+// NotInteger Error
+type NotInteger struct{}
+
+func (e NotInteger) Error() string {
+	return "ID provided is not Integer"
+}
+
+// TokenExpired Error
+type TokenExpired struct {
```
Comment (zaletniy) [inline client.go]: This error should be handled within the client.
If token expired, new one should be issued.
So if you needed it just remane to tokenExpiredError (lower case)

File: client.go [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+// VisualizationClient client for Visualization
+type VisualizationClient struct {
+	url            string
+	client         *http.Client
+	token          AuthToken
+	openstackToken string
+}
+
+// NewVisualizationClient returns client with token
+func NewVisualizationClient(url string, token AuthToken) (*VisualizationClient, error) {
```
Comment (zaletniy) [inline client.go]: it is not AuthToken model, but openstack's one.
Means just string or []byte (i prefer string)

File: client.go [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+type JSONMessage struct {
+	Message string `json:"message"`
+}
+
+// httpRequest handle the request to Grafana server.
+//It returns the response body and a error if something went wrong
+func (v *VisualizationClient) httpRequest(method string, url string, body io.Reader) (result io.Reader, err error) {
+
+	// validate token
+	if v.token == (AuthToken{}) {
```
Comment (zaletniy) [inline client.go]: Move token validation logic one level up.
f.e.
- GetUsers // json models here
   - authorizedHttpRequest // token checking and reissuing new one here
      - httpRequest // just http call here

File: client.go [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+	}
+	tokenExpires := v.token.Token.ExpiresAt.UnixNano() / 1000000
+	now := time.Now().UnixNano() / 1000000
+	if tokenExpires < now {
+		return result, TokenExpired{}
+	}
+
+	request, err := http.NewRequest(method, url, body)
+	request.Header.Set("Content-Type", "application/json")
+	if strings.Contains(url, "openstack") {
```
Comment (zaletniy) [inline client.go]: Too tricky. If you need it. Introduce function parameter `withOpenStackToken`

File: client.go [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+		request.Header.Add("X-OpenStack-Auth-Token", v.openstackToken)
+	} else {
+		bearer := fmt.Sprintf("Bearer %v", v.token)
+		request.Header.Add("Authorization", bearer)
+	}
+
+	response, err := v.client.Do(request)
+	if err != nil {
```
Comment (zaletniy) [inline client.go]: Add original error to error model

File: client.go [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+		return result, VisualizationError{"0", "Unable to perform the http request", "request failed"}
+	}
+
+	//	defer response.Body.Close()
+	if response.StatusCode != 200 {
+		dec := json.NewDecoder(response.Body)
+		var Mess JSONMessage
+		dec.Decode(&Mess)
+
+		return result, VisualizationError{strconv.Itoa(response.StatusCode), Mess.Message, "response not 200"}
```
Comment (zaletniy) [inline client.go]: TODO: discuss API error mapping to client's errors

File: client.go [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+	dec := json.NewDecoder(response)
+	err = dec.Decode(&user)
+
+	return
+}
+
+// CreateUser creates a user
+func (v *VisualizationClient) CreateUser(user User) (userDetails User, err error) {
+	reqURL := v.url + "/admin/users"
+	if len(user.Email) == 0 {
```
Comment (zaletniy) [inline client.go]: just remove all these validations

File: client.go [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+
+	dec := json.NewDecoder(response)
+	err = dec.Decode(&org)
+
+	return
+}
+
+// GetOrganizationID Get Org by ID
+func (v *VisualizationClient) GetOrganizationID(OrgID string) (orgID OrgList, err error) {
+	_, err = strconv.Atoi(OrgID)
```
Comment (zaletniy) [inline client.go]: It shouldn't be an integer. According to spec it is string.

File: client.go [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+
+	dec := json.NewDecoder(response)
+	err = dec.Decode(&orgID)
+
+	return
+}
+
+// DeleteOrganization Delete the organization with given id
+func (v *VisualizationClient) DeleteOrganization(ID string) (org OrgList, err error) {
+	_, err = strconv.Atoi(ID)
```
Comment (zaletniy) [inline client.go]: The same as above

File: client.go [hunk 17]
```diff
@@ -0,0 +1,436 @@
...
+	}
+
+	dec := json.NewDecoder(response)
+	err = dec.Decode(&org)
+
+	return
+}
+
+// CreateOrganization creates a organization
+func (v *VisualizationClient) CreateOrganization(org OrgList) (orgID OrgList, err error) {
```
Comment (zaletniy) [inline client.go]: return value is not orgID, but org model. Naming is confusing

File: client.go [hunk 18]
```diff
@@ -0,0 +1,436 @@
...
+	}
+
+	dec := json.NewDecoder(response)
+	err = dec.Decode(&org)
+	return
+}
+
+// DeleteOrganizationUser Delete User in Organisation
+func (v *VisualizationClient) DeleteOrganizationUser(userID string, orgID string) (org OrgUserList, err error) {
+	_, err = strconv.Atoi(userID)
```
Comment (zaletniy) [inline client.go]: Just remove all validation

**Comment by kbhonagiri16** [comment]
@zaletniy done

File: client.go (hunk 9 above)
Comment (kbhonagiri16) [inline client.go]: openstack token is string and token we get from auth is this model.

File: client.go [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+// httpRequest handle the request to Grafana server.
+//It returns the response body and a error if something went wrong
+func (v *VisualizationClient) httpRequest(method string, url string, body io.Reader) (result io.Reader, err error) {
+	err = v.authorizedHttpRequest()
+	if err != nil {
+		return
+	}
+	request, err := http.NewRequest(method, url, body)
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Add("X-OpenStack-Auth-Token", v.openstackToken)
```
Comment (zaletniy) [inline client.go]: Openstack token header should be sent only on auth request.

File: client.go [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+	response, err := v.client.Do(request)
+	if err != nil {
+		return result, VisualizationError{"0", "Unable to perform the http request", "request failed"}
+	}
+
+	//	defer response.Body.Close()
+	if response.StatusCode != 200 {
+		dec := json.NewDecoder(response.Body)
+		var Mess JSONMessage
+		dec.Decode(&Mess)
```
Comment (zaletniy) [inline client.go]: Please check for error if any during response decoding.

File: client.go [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+
+	response, err := v.client.Do(request)
+	if err != nil {
+		return result, VisualizationError{"0", "Unable to perform the http request", "request failed"}
+	}
+
+	//	defer response.Body.Close()
+	if response.StatusCode != 200 {
+		dec := json.NewDecoder(response.Body)
+		var Mess JSONMessage
```
Comment (zaletniy) [inline client.go]: If status code is not 200, the response is Error model itself, so just decode VisualizationError model

File: client.go (hunk 9 above)
Comment (zaletniy) [inline client.go]: 1. Token should be string.
2. Please add http.Client argument. If it is nil, create your own.

File: client.go [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+type User struct {
+	ID       string `json:"ID"`
+	Email    string `json:"email"`
+	Name     string `json:"name"`
+	Login    string `json:"login"`
+	Password string `json:"password"`
+}
+
+// OpenstackAuth gets a openstack token
+func (v *VisualizationClient) OpenstackAuth() (token AuthToken, err error) {
```
Comment (zaletniy) [inline client.go]: Please hide this method and introduce `func (v *VisualizationClient)Authenticate() (error)` instead

File: client.go [hunk 23]
```diff
@@ -0,0 +1,331 @@
...
+	_, err := v.OpenstackAuth()
+	return err == nil
+}
+
+// authorizedHttpRequest for token checking and reauth
+func (v *VisualizationClient) authorizedHttpRequest() (err error) {
+	// validate token
+	if v.token == (AuthToken{}) {
+		return NotFound{}
+	}
```
Comment (zaletniy) [inline client.go]: No reason for handling it here. If there is no token, we have issue it.

File: client.go [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+	tokenExpires := v.token.Token.ExpiresAt.UnixNano() / 1000000
+	now := time.Now().UnixNano() / 1000000
+	if tokenExpires < now {
+		if v.reIssue() {
+			return v.authorizedHttpRequest()
```
Comment (zaletniy) [inline client.go]: Looks like recursion. Do we really need it here?

File: client.go [hunk 25]
```diff
@@ -0,0 +1,331 @@
...
+
+// NewVisualizationClient returns client with token
+func NewVisualizationClient(url string, token AuthToken) (*VisualizationClient, error) {
+	client := http.Client{}
+	return &VisualizationClient{client: &client, url: url, token: token}, nil
+
+}
+
+// A JSONMessage contains the json error message received when http request failed
+type JSONMessage struct {
```
Comment (zaletniy) [inline client.go]: Not sure why we need it.

File: client.go [hunk 26]
```diff
@@ -0,0 +1,331 @@
...
+	Message string `json:"message"`
+}
+
+func (v *VisualizationClient) reIssue() bool {
+	_, err := v.OpenstackAuth()
+	return err == nil
+}
+
+// authorizedHttpRequest for token checking and reauth
+func (v *VisualizationClient) authorizedHttpRequest() (err error) {
```
Comment (zaletniy) [inline client.go]: Function called authorizedHttpRequest, however it does no http request actually.

File: client.go [hunk 27]
```diff
@@ -0,0 +1,331 @@
...
+	request.Header.Add("X-OpenStack-Auth-Token", v.openstackToken)
+	bearer := fmt.Sprintf("Bearer %v", v.token)
+	request.Header.Add("Authorization", bearer)
+
+	response, err := v.client.Do(request)
+	if err != nil {
+		return result, VisualizationError{"0", "Unable to perform the http request", "request failed"}
+	}
+
+	//	defer response.Body.Close()
```
Comment (zaletniy) [inline client.go]: Yes, response.Body should be closed. Please do so.

**Comment by kbhonagiri16** [comment]
@zaletniy Done acccording to comments. lmk if you need any corrections. thanks

**Comment by zaletniy** [comment]
I think we can go with it :+1:

**Pull request merged**